### PR TITLE
FileStore: Collect device partition information

### PR DESCRIPTION
**Pull request by joehandzik**
Use blkid APIs to get device parition information into the metadata bundle.

Signed-off-by: Joe Handzik <joseph.t.handzik@hp.com>

### CODE REVIEW

**Comment by joehandzik** [comment]
Hey all,

I tested this myself on Centos 7, but non-Centos/Rhel eyeballs would be appreciated. Another thing that I may be best off doing on this pull request would be to add stubs in the configure script in the event that any of the blkid APIs I used are not present. This is part of a larger effort, along with the rest of the Calamari guys (Gregory, Dan, etc), to expose underlying device data into the CLI and REST API. Initially, we'll target storage devices.

**Comment by dachary** [comment]
conditional compilation based on blkid availability sounds like a good idea. As well as a few tests to verify it works as intended.

**Comment by yuyuyu101** [comment]
I'd like we move these things to another helper file. Other backend or else may also can use this :-)

**Comment by joehandzik** [comment]
I thought about that too. Would you want it moved into something like OSDMap? I didn't think that would be the right place, seemed more about aggregation than querying. I wasn't sure that all other storage devices would benefit from the same query, but if you think enough other would...



On May 16, 2015, at 4:32 AM, Haomai Wang <notifications@github.com<mailto:notifications@github.com>> wrote:


I'd like we move these things to another helper file. Other backend or else may also can use this :-)

�

**Comment by joehandzik** [comment]
So, just added what I believe is the right change to deal with the calls I use being absent from libblkid. However, not sure I handled the header file bit right. What environment would I seek out that doesn't actually have blkid.h? If blkid.h isn't present, what would we like to see happen? 

Wasn't quite clear on what tests should be created either, I didn't see anything that tested something similar. Are you just looking for some sort of test in src/test/filestore? Or am I missing the point?

**Comment by yuyuyu101** [comment]
Related unittest is at test/objectstore/store_test.cc

File: src/os/FileStore.cc [hunk 1]
```diff
@@ -641,10 +644,25 @@ bool parse_attrname(char **name)
 
 void FileStore::collect_metadata(map<string,string> *pm)
 {
+  char uuid_str[UUID_LEN+1];
+  blkid_cache cache = NULL;
+  blkid_dev dev = NULL;
+  uuid_d filestore_fsid = get_fsid();
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: could we trust this fsid ? since fsid is read from config file.

Comment (joehandzik) [inline src/os/FileStore.cc]: As much as we can trust any of the others in the code, as far as I can tell. :)

Basically, here's how it works:
1. if you use ceph-disk, the 'fsid' file that gets pushed down is the partition UUID for the underlying filesystem. That's our ideal scenario. Use ceph tools and you'll be able to manage things as expected.
2. if you don't use ceph-disk, you can create an OSD with a UUID that is completely independent from the partition UUID. Unfortunately, without popping out to a script and having more data about my original underlying device, I don't have a C/C++ mechanism to query the data I need. 

I think this is a reasonable compromise. Use ceph-disk/ceph-deploy, and everything will work as expected. Don't, and it's up to you to get the correct metadata in place.

Comment (joehandzik) [inline src/os/FileStore.cc]: FYI, I experimentally tested both of those options myself to confirm them.

**Comment by joehandzik** [comment]
So, I should update the chunk of that test here:

TEST_P(StoreTest, collect_metadata) {
  map<string,string> pm;
  store->collect_metadata(&pm);
  if (GetParam() == string("filestore")) {
    ASSERT_NE(pm.count("filestore_backend"), 0u);
    ASSERT_NE(pm.count("filestore_f_type"), 0u);
  }
}

Right? I'll get that done today. Please let me know if there's anything else I need to get done to have this patch (er, patchset at this point) ready to go up. Thanks for the pointer!

**Comment by yuyuyu101** [comment]
@liewegas what do you think about this change?

File: configure.ac [hunk 2]
```diff
@@ -262,11 +262,19 @@ ACX_PTHREAD
...
 
+# libblkid
+AC_CHECK_HEADER([blkid/blkid.h], [],
+  AC_MSG_ERROR([blkid/blkid.h not found (libblkid-dev, libblkid-devel)]))
+AC_CHECK_LIB([blkid], [blkid_get_cache], [true],
+  AC_MSG_FAILURE([libblkid not found]))
+AC_CHECK_LIB([blkid], [blkid_find_dev_with_tag], [true],
+  AC_MSG_FAILURE([libblkid not found]))
+AC_CHECK_LIB([blkid], [blkid_dev_devname], [true],
+  AC_MSG_FAILURE([libblkid not found]))
```
Comment (liewegas) [inline configure.ac]: This still needs to be guarded by the linux check to avoid breaking other OSs.

File: src/os/FileStore.cc [hunk 3]
```diff
@@ -641,10 +644,25 @@ bool parse_attrname(char **name)
...
+
+  uuid_unparse((const unsigned char*)&filestore_fsid.uuid, uuid_str);
+
+  if (blkid_get_cache(&cache, NULL) >= 0)
+    dev = blkid_find_dev_with_tag(cache, "PARTUUID", (const char*)uuid_str);
+
+  if (dev)
+    (*pm)["filestore_partition_path"] = string(blkid_dev_devname(dev));
+  else
+    (*pm)["filestore_partition_path"] = "unknown";
```
Comment (liewegas) [inline src/os/FileStore.cc]: I think that *both* the data and journal partitions will be tagged with the osd UUID, just with different partition types.  We should include both in the metadata?

**Comment by liewegas** [comment]
I'm a little sad that this is a bottom-up approach (assume UUID stuff is correct and work backwards) vs a top-down approach (look at the path, identify the mount point, identify the associated block devices), but the latter would be way more work to implement.  This is so simple it is a no-brainer.

It would be nice to put this in a helper in common/blk* or something.  Say,

 int get_device_by_uuid(uuid_d type, uuid_d label, string *device);

?  Then there's just one bit of code that needs the conditional #include (linux only) etc.

File: configure.ac (hunk 2 above)
Comment (joehandzik) [inline configure.ac]: I can do that, but would I need to move the code into a linux-only compiled object (i.e., not FileStore), right? I suppose I can take care of that via your other suggestion down below.

**Comment by joehandzik** [comment]
Fair enough, I'll move it out of FileStore.

I just didn't see the added value in going the top down direction. Sure, it would make up for the UUID not mapping to the partition UUID, but I think we should enforce that. I had also considered adding code to Ceph disk to make a new, separate attribute of some sort (probably yank the partition UUID in case it doesn't match what I'm currently querying). Would that be preferable? My take is that we can change it later if necessary, but what I have here is fine for now.

Thanks for taking a look!

Joe

On May 27, 2015, at 12:01 PM, Sage Weil <notifications@github.com<mailto:notifications@github.com>> wrote:


I'm a little sad that this is a bottom-up approach (assume UUID stuff is correct and work backwards) vs a top-down approach (look at the path, identify the mount point, identify the associated block devices), but the latter would be way more work to implement. This is so simple it is a no-brainer.

It would be nice to put this in a helper in common/blk* or something. Say,

int get_device_by_uuid(uuid_d type, uuid_d label, string *device);

? Then there's just one bit of code that needs the conditional #include (linux only) etc.

�

File: src/os/FileStore.cc (hunk 3 above)
Comment (joehandzik) [inline src/os/FileStore.cc]: I just did a quick test and did not get the same UUID for the journal and data disks. So I'd need to add a separate call out to grab the journal partition path too, but I'm not sure the UUID I need is in the code currently. I'll take a look, but my gut tells me that'll be a separate patch.

**Comment by liewegas** [comment]
On Wed, 27 May 2015, Joe Handzik wrote:
> Fair enough, I'll move it out of FileStore.

Thanks!

> I just didn't see the added value in going the top down direction. Sure, it would make up for the UUID not mapping to the partition UUID, but I
> think we should enforce that. I had also considered adding code to Ceph disk to make a new, separate attribute of some sort (probably yank the
> partition UUID in case it doesn't match what I'm currently querying). Would that be preferable? My take is that we can change it later if
> necessary, but what I have here is fine for now.

Yeah, I agree this approach is simpler and totally adequate.  I was 
getting paranoid about non-standard deployments but I don't think it's 
worth worry about those.

**Comment by liewegas** [comment]
Needs rebase!

**Comment by joehandzik** [comment]
Huh, not entirely sure what went wrong on the failure. I'll look into it more tomorrow when I pull in my updated changes. A side benefit to moving into the common/blkdev.cc is that I have easy access to a parsing function, so I'm pulling the base device out alongside the partition path. One less thing I'll need to do in Calamari!

**Comment by joehandzik** [comment]
Ok, I think this is finally, actually done. Lemme know if you have any issues with the formatting, tested on Rhel7 and the unit test passes. At some point I must have squashed my unit test patch with another patch, I can squash everything together if that's better or try to figure out a way to separate the unit test patch out again.

**Comment by joehandzik** [comment]
Looking at this failure now

**Comment by joehandzik** [comment]
Looks like it failed because of a missing header file in one of the tests. The failure actually makes sense, but what doesn't make sense is that this was building fine on my machine. Not sure how that works. Anyway, this should fix it, I ran into the same error while testing my build early-on. If this doesn't fix it, I'll pull down something fresh (guess my current code path has something that doesn't get cleaned out with a 'make clean').

**Comment by joehandzik** [comment]
Should be fixed now, got a clean clone to work. The issues were only present when building with --with-debug. For some reason, the header file order was a problem only in the blkdev test file, but not in the blkdev object itself. Also, found that the blkdev test file was missing the -luuid link that was already around for the blkdev object, fixed that too.

**Comment by joehandzik** [comment]
Hey Ceph team,

Is there any more work you'd like to see put into this PR?

File: src/os/FileStore.cc [hunk 4]
```diff
@@ -641,10 +642,31 @@ bool parse_attrname(char **name)
...
+
+  switch (rc) {
+    case -EINVAL:
+      (*pm)["backend_filestore_partition_path"] = "unknown";
+      (*pm)["backend_filestore_dev_node"] = "unknown";
+      break;
+    case -ENODEV:
+      (*pm)["backend_filestore_partition_path"] = string(partition_path);
+      (*pm)["backend_filestore_dev_node"] = "unknown";
+      break;
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: handle -EOPNOTSUPP?

Comment (joehandzik) [inline src/os/FileStore.cc]: You're right, I missed that. I'll add it in the next couple hours.

Joe



On Jun 10, 2015, at 8:56 AM, Haomai Wang <notifications@github.com<mailto:notifications@github.com>> wrote:


In src/os/FileStore.cc<https://github.com/ceph/ceph/pull/4699#discussion_r32119162>:

>    (*pm)["filestore_backend"] = backend->get_name();
>    ostringstream ss;
>    ss << "0x" << std::hex << m_fs_type << std::dec;
>    (*pm)["filestore_f_type"] = ss.str();
> +
> +  rc = get_device_by_uuid(get_fsid(), "PARTUUID", partition_path,
> +        dev_node);
> +
> +  switch (rc) {
> +    case -EINVAL:
> +      (*pm)["backend_filestore_partition_path"] = "unknown";
> +      (*pm)["backend_filestore_dev_node"] = "unknown";
> +      break;
> +    case -ENODEV:
> +      (*pm)["backend_filestore_partition_path"] = string(partition_path);
> +      (*pm)["backend_filestore_dev_node"] = "unknown";
> +      break;


handle -EOPNOTSUPP?

�

**Comment by joehandzik** [comment]
Based on the number of builds with the exact same issue on their console, I'm assuming this isn't a problem I have any control over...

**Comment by dachary** [comment]
<pre>
./common/blkdev.h:12:31: warning: 'get_device_by_uuid' initialized and declared 'extern' [enabled by default]
 extern int get_device_by_uuid(uuid_d dev_uuid, const char* label,
                               ^
./common/blkdev.h:12:31: error: 'uuid_d' was not declared in this scope
./common/blkdev.h:12:48: error: expected primary-expression before 'const'
 extern int get_device_by_uuid(uuid_d dev_uuid, const char* label,
                                                ^
./common/blkdev.h:13:3: error: expected primary-expression before 'char'
   char* partition, char* device);
   ^
./common/blkdev.h:13:20: error: expected primary-expression before 'char'
   char* partition, char* device);
                    ^
./common/blkdev.h:13:32: error: expression list treated as compound expression in initializer [-fpermissive]
   char* partition, char* device);
                                ^
</pre>
it looks like a compilation issue related to your commit.

**Comment by joehandzik** [comment]
You're right, pushed from an old directory and hadn't run configured with the right optional flags. I take it that this (http://jenkins.ceph.dachary.org/job/ceph/5797/console) is no longer a good place to go and look for errors? Would explain why all other failed builds looked the same if that's just a place to see that there was some sort of error.

**Comment by dachary** [comment]
http://jenkins.ceph.dachary.org/job/ceph/5797/console is a reliable way to know if your pull request works or not.

**Comment by dachary** [comment]
@joehandzik I misunderstood your question, sorry about that. http://jenkins.ceph.dachary.org/job/ceph/5797/console shows the agregation of all the runs for your pull request. From there you find links to the centos and ubuntu builds. It's indeed not a good place to look for errors, it only shows the global status of the run.

**Comment by joehandzik** [comment]
@dachary No worries! Ok, I'll keep that in mind next time, just got myself in the habit of being able to look at that single location to identify build issues. All should be building fine now, and hopefully we'll be able to finish this PR up...

**Comment by yuyuyu101** [comment]
LGTM except a little sad fsid only useful for strict deploy way.

**Comment by joehandzik** [comment]
Well, lemme make sure this is better explained...

ceph-deploy just happens to be a standard tool that executes the steps needed to track an OSD and a filesystem partition by the same UUID. If you don't want to use ceph-deploy, you can:

1. Create filesystem partition.
2. Take note of the UUID generated (probably easiest via blkid)
3. When creating the OSD, pass in the filesystem UUID as the OSD UUID.

This will provide the OSD a sane, useful UUID. There may be other ways to get to this point, but this is probably the easiest way (again, without using ceph-deploy).

**Comment by joehandzik** [comment]
@yuyuyu101 ,

How do I move forward from here? Do I need to add you with a "Reviewed-By" tag? Or do I still wait for Sage to give input?

**Comment by liewegas** [comment]
This looks good ot me.  Adding needs-qa tag so it takes a pass through the test suite before merge

**Comment by joehandzik** [comment]
Ok, thanks Sage!

**Pull request merged**